micrometer-core/src/main/java/io/micrometer/core/instrument/distribution/FixedBoundaryHistogram.java

```diff
@@ -73,28 +93,49 @@ else if (value > key)
         return low < buckets.length ? low : -1;
     }
 
-    Iterator<CountAtBucket> countsAtValues(Iterator<Double> values) {
+    Iterator<CountAtBucket> countsAtValues(Iterator<Double> buckets) {
```

Comment:
Can we just use `getCountsAtBucket()` instead of this? This makes a call to `countAtBucket()` for every bucket which in turn does a binary search to find the bucket.  It makes sense if we need to get the counts for only a subset of buckets, but we don't do that anywhere (why would we even do that?). But at the end, all we need is a continuous list of buckets and their count which we can get by iterating on the values directly.

There were a few issues solved around this area hence not changing anything at this point. But, 
my claim would be the `buckets` would not change after the histogram is created so we should always rely on the buckets from the underlying implementation (in this case FixedBucketHistogram).

Reply:
> It makes sense if we need to get the counts for only a subset of buckets, but we don't do that anywhere (why would we even do that?).

I'm not sure why we would but we did at some point (inadvertently) do that, which caused a regression when code in this area was changed - see https://github.com/micrometer-metrics/micrometer/commit/bbb5585e60cc8c7aee481b3e90a1429196e58427. Unless we find some reason to do that, I agree we don't have a use case for it now. I'd be willing to explore a refactor that gets rid of this, but not in maintenance branches, I think.